driver/labjack/reader.h

```diff
@@ -0,0 +1,363 @@
+    std::string serial_number; // used to open devices
+    ///@brief The type of connection (i.e. USB, Ethernet, or WIFI)
+    std::string connection_type; // used to open devices
+    ///@brief map of locations on device to synnax channel keys
+    std::map<std::string, uint32_t> channel_map;
+    std::vector<std::string> phys_channels;
+    std::vector<ReaderChannelConfig> tc_channels;
+    std::set<uint32_t> tc_index_keys;
+    ///@brief whether to persist data to disk
+    bool data_saving;
+
+    ReaderConfig() = default;
+
+    explicit ReaderConfig(config::Parser &parser)
+        : device_type(parser.optional<std::string>("type", "")),
+          device_key(parser.required<std::string>("device")),
+          sample_rate(synnax::Rate(parser.optional<int>("sample_rate", 1))),
+          stream_rate(synnax::Rate(parser.optional<int>("stream_rate", 1))),
+          serial_number(parser.required<std::string>("device")),
+          connection_type(parser.optional<std::string>("connection_type", "")),
+          data_saving(parser.optional<bool>("data_saving", false)
+          ) {
+        if (!parser.ok())
+            LOG(ERROR) << "Failed to parse reader channel config: " << parser.
+                    error_json().dump(4);
+
+        parser.iter("channels", [this](config::Parser &channel_parser) {
+            auto channel = labjack::ReaderChannelConfig(channel_parser);
+
+            if (channel.enabled && channel.channel_type != "TC") {
+                channels.emplace_back(channel);
+                this->phys_channels.push_back(channel.location);
+            } else if (channel.enabled && channel.channel_type == "TC") {
+                tc_channels.emplace_back(channel);
+            }
+
+            this->channel_map[channel.location] = channel.key;
+        });
+    }
+};
+
+///////////////////////////////////////////////////////////////////////////////////
+//                                   ReaderSource                                //
+///////////////////////////////////////////////////////////////////////////////////
+class ReaderSource : public pipeline::Source {
+public:
+    explicit ReaderSource(
+        const std::shared_ptr<task::Context> &ctx,
+        const synnax::Task task,
+        const ReaderConfig &reader_config
+    ) : ctx(ctx),
+        task(task),
+        reader_config(reader_config) {
+        auto breaker_config = breaker::Config{
+            .name = task.name,
+            .base_interval = 1 * SECOND,
+            .max_retries = 20,
+            .scale = 1.2,
+        };
+        this->breaker = breaker::Breaker(breaker_config);
+    }
+
+    ~ReaderSource();
+
+    std::vector<synnax::ChannelKey> get_channel_keys();
+
+    void stopped_with_err(const freighter::Error &err);
+
+    std::pair<Frame, freighter::Error> read(breaker::Breaker &breaker);
+
+    freighter::Error stop(const std::string &cmd_key);
+
+    freighter::Error start(const std::string &cmd_key);
+
+    void stop();
```

Comment:
Clion is telling me this method isn't implemented. Is that true?

Reply:
yes. removing